api/src/routes/donate.ts

```diff
@@ -214,3 +218,156 @@ export const donateRoutes: FastifyPluginCallbackTypebox = (
+          stripeSubscription.latest_invoice.payment_intent &&
+          typeof stripeSubscription.latest_invoice.payment_intent !==
+            'string' &&
+          stripeSubscription.latest_invoice.payment_intent.client_secret !==
+            null
+        ) {
+          const clientSecret =
+            stripeSubscription.latest_invoice.payment_intent.client_secret;
+          return reply.send({
+            subscriptionId: stripeSubscription.id,
+            clientSecret
+          });
+        } else {
+          throw new Error('Stripe payment intent client secret is missing');
+        }
+      } catch (error) {
+        fastify.log.error(error);
+        fastify.Sentry.captureException(error);
+        void reply.code(500);
+        return reply.send({
+          error: 'Donation failed due to a server error.'
+        });
+      }
+    }
+  );
+
+  fastify.post(
+    '/donate/charge-stripe',
+    {
+      schema: schemas.chargeStripe
+    },
+    async (req, reply) => {
+      try {
+        const { email, amount, duration, subscriptionId } = req.body;
+
+        const subscription =
+          await stripe.subscriptions.retrieve(subscriptionId);
+        const isSubscriptionActive = subscription.status === 'active';
+        const productId = subscription.items.data[0]?.plan.product?.toString();
+        const isStartedRecently = inLastFiveMinutes(
+          subscription.current_period_start
+        );
+        const isProductIdValid =
+          productId && allStripeProductIdsArray.includes(productId);
+
+        const isValidCustomer = typeof subscription.customer === 'string';
+        if (
+          isSubscriptionActive &&
+          isProductIdValid &&
+          isStartedRecently &&
+          isValidCustomer
+        ) {
+          // TODO(Post-MVP) new users should not be created if user is not found
+          const user = await findOrCreateUser(fastify, email);
+          const donation = {
+            userId: user.id,
+            email,
+            amount,
+            duration,
+            provider: 'stripe',
+            subscriptionId,
+            customerId: subscription.customer as string,
+            // TODO(Post-MVP) migrate to startDate: new Date()
+            startDate: {
+              date: new Date().toISOString(),
+              when: new Date().toISOString().replace(/.$/, '+00:00')
+            }
+          };
+
+          await fastify.prisma.donation.create({
+            data: donation
+          });
+
+          await fastify.prisma.user.update({
+            where: { id: user.id },
+            data: {
+              isDonating: true
+            }
+          });
+          return reply.send({
+            isDonating: true
+          });
+        } else {
+          throw new Error('Stripe subscription information is invalid.');
+        }
+      } catch (error) {
+        fastify.log.error(error);
+        fastify.Sentry.captureException(error);
```

Comment:
I would include more information in the error. Currently, in sentry, all you are going to get is `"Stripe subscription information is invalid"`. But, you will not know what/why.

So, whatever information you think is important, I would add. Maybe everything in the `if` condition?

Reply:
Thanks for the review, I added db checks and more descriptive error messages.